### PR TITLE
refactor(core): split `ɵɵdefineDirective` and `ɵɵdefineComponent`

### PR DESCRIPTION
**Pull request by alan-agius4**
Before this change `ɵɵdefineDirective` called `ɵɵdefineComponent` under the hood. This is problematic for the consistent component id generation as it could result in hash collisions for certain directives. Directives however do not require an id.

This changes moves common definition generation logic into a separate function that is re-used in `ɵɵdefineDirective` and `ɵɵdefineComponent`.

### CODE REVIEW

File: packages/core/src/render3/definition.ts [hunk 1]
```diff
@@ -175,13 +144,43 @@ export function ɵɵdefineComponent<T>(componentDefinition: {
...
+}
+
+interface ComponentDefinition<T> extends Omit<DirectiveDefinition<T>, 'features'> {
+  /**
+   * The number of nodes, local refs, and pipes in this component template.
+   *
+   * Used to calculate the length of this component's LView array, so we
+   * can pre-fill the array and set the binding start index.
+   */
+  // TODO(kara): remove queries from this count
```
Comment (AndrewKushnir) [inline packages/core/src/render3/definition.ts:173]: We can drop this TODO, it's unlikely that we'll change this anytime soon.
```suggestion
```

File: packages/core/src/render3/definition.ts [hunk 2]
```diff
@@ -543,143 +528,21 @@ function invertObject<T>(
...
-    ɵɵdefineComponent as any as<T>(directiveDefinition: {
-      /**
-       * Directive type, needed to configure the injector.
-       */
-      type: Type<T>;
-
-      /** The selectors that will be used to match nodes to this directive. */
-      selectors?: CssSelectorList;
-
-      /**
```
Comment (AndrewKushnir) [inline packages/core/src/render3/definition.ts:556]: It looks like we've lost some of these comments. Let's carry them over to the new locations as well.

**Review by AndrewKushnir — COMMENTED** [review]
@alan-agius4 looks great 👍 Just left some minor comments.

(we should probably also run a TGP to validate the change)

File: packages/core/src/render3/definition.ts (hunk 2 above)
Comment (alan-agius4) [reply packages/core/src/render3/definition.ts:556]: The comments have not been lost. This is due to removing the entire interfaces which was duplicate.

File: packages/core/src/render3/definition.ts [hunk 3]
```diff
@@ -543,143 +528,21 @@ function invertObject<T>(
...
-      exportAs?: string[];
-    }) => never;
+export function ɵɵdefineDirective<T>(directiveDefinition: DirectiveDefinition<T>):
+    Mutable<DirectiveDef<any>, keyof DirectiveDef<any>> {
+  return noSideEffects(() => {
+    const def = getNgDirectiveDef(directiveDefinition);
+
+    // TODO(alanagius): the below should not be needed but there are a number of checks
+    // that don't handle 'undefined' template for directives.
+    (def as any).template = null;
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply packages/core/src/render3/definition.ts]: Looks like there were only 2 after all, just took some time to find the last one. 

- https://github.com/angular/angular/pull/49350/files#diff-26a0112e0e31c4ce31c9fc0eb883f7fcab197f282e4f84eb2f6ff59e46bb8f9aR45
- https://github.com/angular/angular/pull/49350/files#diff-c57c65294ada4e4371086f3e395b0e1920ce087b9c9fa8a8620e8c33fe0d29ddR443

**Comment by alan-agius4** [comment]
TGP is green. 
http://test/OCL:514687509:BASE:514996711:1678282386002:6c8af224

**Comment by AndrewKushnir** [comment]
@alan-agius4 could you please rebase this PR when you get a chance? It looks like it has a merge conflict with the main branch at this moment.

**Comment by alan-agius4** [comment]
@AndrewKushnir rebased.

File: packages/core/src/render3/definition.ts [hunk 4]
```diff
@@ -276,74 +263,61 @@ export function ɵɵdefineComponent<T>(componentDefinition: {
...
-        null;
-    def.pipeDefs = dependencies ?
-        (() => (typeof dependencies === 'function' ? dependencies() : dependencies)
-                   .map(getPipeDef)
-                   .filter(nonNull)) :
-        null;
 
+    initFeatures(def);
+    const dependencies = componentDefinition.dependencies;
+    def.directiveDefs = extractDefListOrFactory(dependencies, /* pipeDef */ false);
```
Comment (AndrewKushnir) [inline packages/core/src/render3/definition.ts:319]: nit: can we pass the def extractor from here (vs passing a flag)?

(**not** a blocker for this PR, we can do in a followup)

File: packages/core/src/render3/util/discovery_utils.ts [hunk 5]
```diff
@@ -440,7 +440,7 @@ function sortListeners(a: Listener, b: Listener) {
  * See call site for more info.
  */
 function isDirectiveDefHack(obj: any): obj is DirectiveDef<any> {
-  return obj.type !== undefined && obj.template !== undefined && obj.declaredInputs !== undefined;
+  return obj.type !== undefined && obj.declaredInputs !== undefined;
```
Comment (AndrewKushnir) [inline packages/core/src/render3/util/discovery_utils.ts:443]: nit: we could also may be replace the `obj.template` with smth else like `obj.findHostDirectiveDefs !== undefined` (which would be true for both directives and components), so we have a better check?

**Review by AndrewKushnir — APPROVED** [review]
@alan-agius4 looks great 👍 Just a couple minor comments that we can address in a followup PR if needed (so that we don't invalidate the TGP run).

File: packages/core/src/render3/definition.ts (hunk 4 above)
Comment (alan-agius4) [reply packages/core/src/render3/definition.ts:319]: Let me do that in a follow up.

File: packages/core/src/render3/util/discovery_utils.ts (hunk 5 above)
Comment (alan-agius4) [reply packages/core/src/render3/util/discovery_utils.ts]: Done.

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit 852558366d2a5e8c7b2f7d9f01e8c328d3b54f85.

**Pull request closed without merging**